### PR TITLE
Major Refactor

### PR DESCRIPTION
**Pull request by ariG23498**
@ayulockin could you please review this?

TODOs after review:
- [x] Run the baseline and gather the wandb and hf artefacts
- [ ] Create an inference pipeline
- [ ] Showcase the model with a demo

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,25 +1,44 @@
-# LLM Deep Dive
+# Attention Is All You Need - Educational Repository
```
Comment (ayulockin) [inline README.md:1]: I still think it should be "LLM Deep Dive - Educational Repository". Here's my reasoning:

- the repo is meant for us to deeply learn about the whole LLM fiasco with implementation
- we are not only doing "Attention is all you need"
- we will eventually move to Vision/Video and can change the name then.

File: README.md [hunk 2]
```diff
@@ -1,25 +1,44 @@
...
 
-Exploring LLMs by implementing and understanding all the key components, perform experiments and optimize it.
+Welcome to the **Educational** Attention Is All You Need repository!
```
Comment (ayulockin) [inline README.md:3]: same.

File: baseline/README.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+- `01_build_dataset.py` - Script to build the dataset for training.
+- `02_build_tokenizer.py` - Script to create a tokenizer.
+- `03_training.py` - Main script to train the baseline Transformer model.
+- `config.py` - Contains configuration details such as model hyperparameters and paths.
+- `requirements.txt` - Lists the Python dependencies required to run the code.
+
+
+## Getting Started
+
```
Comment (ayulockin) [inline baseline/README.md:28]: Fork this repository followed by:

File: baseline/README.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+```bash
+git clone https://github.com/your-username/attention-is-all-you-need.git
+cd attention-is-all-you-need/baseline
```
Comment (ayulockin) [inline baseline/README.md:31]: Let's keep the repo name as `llm_scratch` for now.

File: baseline/02_build_tokenizer.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+
+def batch_iterator(dataset, batch_size=100_000):
+    for sample in dataset.iter(batch_size):
+        texts = sample["en"] + sample["de"]
+        yield texts
+
+
+if __name__ == "__main__":
+    dataset = load_dataset(config.DE_EN_SPLIT_DATASET, split="train+test+validation")
```
Comment (ayulockin) [inline baseline/02_build_tokenizer.py:23]: Should we build the tokenizer using the test and validation set?

**Review by ayulockin — COMMENTED** [review]
I am not done reviewing. Left some comments on the naming and stuff. You can take a look. Realised that the model and training pipeline is in one script so I need to pull in this branch and run things to thoroughly review this PR.

File: baseline/02_build_tokenizer.py (hunk 5 above)
Comment (ariG23498) [reply baseline/02_build_tokenizer.py:23]: I am not quite sure, and there is not mention of how to train the tokenizer in the paper. What do you think?

**Comment by ariG23498** [comment]
@ayulockin ready for another review, running the script and then merging.

File: baseline/03_training.py [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+        dropout_rate=config.DROPOUT_RATE,
+        max_seq_len=config.MAX_SEQ_LENGTH,
+        vocab_size=config.VOCAB_SIZE,
+    )
+
+    total_num_batches = len(train_dataloader)
+    total_steps = total_num_batches * config.NUM_EPOCHS
+    warmup_steps = int(config.WARMUP_PERCENTAGE * total_steps)
+
+    model = Transformer(config).to("cuda")
```
Comment (ayulockin) [inline baseline/03_training.py:478]: you should be passing `model_config`

File: baseline/03_training.py [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+
+    tokenizer = Tokenizer.from_pretrained(config.TOKENIZER_ID)
+    tokenizer.enable_padding(pad_token=config.SPECIAL_TOKENS["pad_token"])
+    tokenizer.enable_truncation(config.MAX_SEQ_LENGTH)
+
+    train_dataloader = DataLoader(
+        dataset=train_dataset,
+        batch_size=config.BATCH_SIZE,
+        shuffle=True,
+        collate_fn=collate_fn,
```
Comment (ayulockin) [inline baseline/03_training.py:454]: `collate_fn=partial(collate_fn, tokenizer=tokenizer),`

File: baseline/03_training.py [hunk 8]
```diff
@@ -0,0 +1,550 @@
...
+    scheduler = LambdaLR(optimizer, lr_lambda=partial(learning_rate_schedule, model_dim=config.MODEL_DIM))
+
+    loss_fn = nn.CrossEntropyLoss()
+
+    run = wandb.init(entity=config.WANDD_ENTITY, project=config.WANDB_PROJECT)
+
+    for epoch in range(config.NUM_EPOCHS):
+
+        # Training Loop
+        for idx, batch in enumerate(train_dataloader):
```
Comment (ayulockin) [inline baseline/03_training.py:493]: Just do something like:

```
progress_bar = tqdm(train_dataloader, desc=f"Epoch {epoch+1} - Training", leave=True)

for idx, batch in enumerate(progress_bar):
```

File: baseline/03_training.py [hunk 9]
```diff
@@ -0,0 +1,550 @@
...
+            run.log({"train-loss": loss.item()})
+
+            loss.backward()
+
+            if idx % config.GRAD_ACCUMULATION_STEP == 0:
+                optimizer.step()
+                optimizer.zero_grad()
+
+            scheduler.step()
+
```
Comment (ayulockin) [inline baseline/03_training.py:519]: With the right indentation do: 

```
progress_bar.set_postfix(loss=loss)
```

Do the same for the valid loop.

**Review by ayulockin — CHANGES_REQUESTED** [review]

**Comment by ayulockin** [comment]
I left a few more comments and here are two notes as well:

- I tried overfitting on one batch and it didn't quite went all the way to 0. W&B run: https://wandb.ai/llm-scratch/wmt14-de-en/runs/7g5oakz9?nw=nwuserayushthakur&panelDisplayName=train-loss
- I see what your model's implementation is not 1:1 with what we have in the `main` branch atm. I did make a few modifications like use of `register_buffer` as in [here](https://github.com/ayulockin/llm_scratch/blob/14c7ca5ddaf27ed13dd6e3968987a708d11ad16c/src/llm/wmt_model.py#L389). I also think the implementation of the attention layer vary (implementation choice) slightly.

File: baseline/requirements.txt [hunk 10]
```diff
@@ -0,0 +1,4 @@
+datasets
+tokenizers
+wandb
+jaxtyping
```
Comment (ayulockin) [inline baseline/requirements.txt:4]: also add torch

File: baseline/03_training.py [hunk 11]
```diff
@@ -0,0 +1,550 @@
...
+                decoder_input=decoder_input,
+                enc_key_attention_mask=enc_key_attention_mask,
+                dec_key_attention_mask=dec_key_attention_mask,
+            )
+
+            logits = logits.view(-1, logits.size(-1))
+            target = batch["labels"][..., 1:].to("cuda").view(-1).long()
+
+            loss = loss_fn(input=logits, target=target)
+            run.log({"train-loss": loss.item()})
```
Comment (ayulockin) [inline baseline/03_training.py:510]: `run.log({"batch/train_loss": loss.item()})`

File: baseline/03_training.py [hunk 12]
```diff
@@ -0,0 +1,550 @@
...
+                    decoder_input=decoder_input,
+                    enc_key_attention_mask=enc_key_attention_mask,
+                    dec_key_attention_mask=dec_key_attention_mask,
+                )
+
+                logits = logits.view(-1, logits.size(-1))
+                target = batch["labels"][..., 1:].to("cuda").view(-1).long()
+
+                val_loss = loss_fn(input=logits, target=target)
+                run.log({"val-loss": val_loss.item()})
```
Comment (ayulockin) [inline baseline/03_training.py:544]: `run.log({"batch/val_loss": loss.item()})`

**Review by ayulockin — CHANGES_REQUESTED** [review]

**Comment by ayulockin** [comment]
Okay I swapped the attention block with the implementation in the `main` branch and we couldn't "overfit": https://wandb.ai/llm-scratch/wmt14-de-en/runs/x7llf2hl/panel/09ollpdh1?nw=nwuserayushthakur

File: baseline/config.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+# Tokenizer
+VOCAB_SIZE = 37_000
+MIN_FREQUENCY = 2
+SPECIAL_TOKENS = {
+    "pad_token": "[PAD]",
+    "begin_token": "[BOS]",
+    "end_token": "[EOS]",
+    "unknown_token": "[UNK]",
+}
+TOKENIZER_ID = "llm-scratch/wmt-14-de-en-tokenizer"
```
[the remark this answers is not in the document]
Comment (ariG23498) [reply baseline/config.py:14]: You are right. I realised that we need to rename the tokenizer at HF to standardise the naming convention. Will do it at HF's end.

**Comment by ariG23498** [comment]
@ayulockin could you give me a reivew now?

Here are certain things that I have changed:
1. Used a tokenizer from Hub (instead of working on it ourselves)
2. Mainly focues on revamping the names of variables
3. Included padding masks inside the loss function

I am still working on the attention implementation, so that it is more aligned to the paper.

**Comment by ayulockin** [comment]
> Used a tokenizer from Hub (instead of working on it ourselves)

@ariG23498 is there a specific reason to do this?

**Comment by ariG23498** [comment]
For the following reasons:
- more stable than our implementation
- easy eos and bos token
- padding and attention mask on the fly
- no change in the api

**Comment by ayulockin** [comment]
Hey @ariG23498 we will probably merge this with the `generate` method from my end added to this PR. 

I am doing this full run: https://wandb.ai/llm-scratch/wmt14-de-en/runs/j14vk5dy/panel/09ollpdh1?nw=nwuserayushthakur

Once done we will do it with the `generate` method too.
